### PR TITLE
introduce random assignment prediction (fixes #5177)

### LINKED ISSUE #5177: Random assignment for a subset of variables in predict.
## Summary

Add random assignment for a subset of variables in predict.

## Motivation

Random assignment for a variable V in a tree is a way to drop the effect of that variable in the model. The procedure is assign randomly in every node that uses V for spliting and in the rest do as usual. 
Random assignment of a subset of variables W = {V1, V2..Vi} is similar. The assignment in nodes which splitting use any variable in W is random and the rest as usual.

This is usefull for computing honest variable importance over a new test set as the difference between the prediction error with and without perturbation.
This method can be used for interaction computation of two subsets of variables V and W, defined as the difference between paired importance (importance when both V and W are assigned randomly) vs added importance (sum of importance when each of W or V is assigned randomly).

## References

The method is referenced in:
Hemant Ishwaran. "Variable importance in binary regression trees and forests." Electron. J. Statist. 1 519 - 537, 2007. https://doi.org/10.1214/07-EJS039

### PR DESCRIPTION
**Pull request by suncanghuai**
* Introduce a struct of parameters PredictionControlParameter

* Add a new field random_assign_features to class Config

* Add new GetLeaf functions that apply random assign mechanism to class Tree

Hi @guolinke, I completed my code writing following a modified implementation plan, would you please give some suggestions for my code? It's my first time creating a PR for LightGBM and any suggestions would be greatly appreciated. Since the implementation plan has been modified from that in the discussion of #5177, there is one thing I should explain before your review:

### Why add a parameter to prediction functions in Predictor/Tree instead of adding some new predict functions?
The difference between normal prediction and random assignment prediction is mainly in how to search a Tree(how to choose to go left or go right), which means they have too much repetitive code except for the code of GetLeaf functions. In the part of predict functions, they have the same logic. That's why I didn't choose to add new predict functions for class Predictor or class Tree. While considering scalability, a structure of control parameters that could be extended easily seems a perfect idea to me.

### CODE REVIEW

File: src/application/predictor.hpp
Comment (jameslamb) [inline src/application/predictor.hpp:44]: What's the benefit of introducing this new `PredictionControlParameter` class instead of just adding a new argument to the `Predictor`, just as we do for other customizations like `predict_leaf_index` and `predict_contrib`?

From my read of the code, it looks like unnecessary extra indirection and complexity, and I'm not sure what it gives us in exchange for that.

File: include/LightGBM/config.h
Comment (jameslamb) [inline include/LightGBM/config.h:866]: Please make it clear in the documentation that this is a vector of feature *indexes* (integers containing the 0-based position of the features in the training data matrix) and not feature *names* (strings corresponding to column names).

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your interest in LightGBM. I'm personally unsure about the value of adding this to LightGBM. I understand the statement from #5177...

> Random assignment for a variable V in a tree is a way to drop the effect of that variable in the model

But what is the value of this being **in LightGBM**, instead of having user code or some other library just produce input with randomly-varied feature values and then using LightGBM's existing `predict()` functions?

We have a very small team of maintainers here with a very small amount of time and effort available relative to the size of the project, and I'm personally not convinced that this new functionality is something we should take into LightGBM.

I'm only one vote though, will defer to @guolinke's and @shiyu1994 's opinion.

-----

If we do move forward with this, I expect to see this PR also add unit tests confirming that the functionality works as expected. Since it's new functionality, I'd expect that none of LightGBM's existing unit tests should need to be modified at all.

I'd like to see tests covering all of the following cases:

* random assignment of one feature
* random assignment of all features
* `random_assign_features` contains feature indices that are too large (e.g. model has 10 features and you pass `random_assign_features=27`)
* `random_assign_features` contains duplicates of the same feature
* `random_assign_features` works as expected with all of LightGBM's prediction types:
    - raw score
    - feature contributions (`pred_contrib`)
    - leaf index (`pred_leaf_index`)

File: src/application/predictor.hpp
Comment (suncanghuai) [reply src/application/predictor.hpp:44]: If only considering random-assignment-prediction, a new parameter to Predictor of vector<int> is enough and more simple. What led me to introduce PredictionControlParameter is that I found there are many scattered code modification for adding a new parameter. I think introducing a structure of parameters could lower the code changes for the next variant of the prediction function. But I have to say that I have also wondered whether it is necessary. So If you guys finally decide to accept this feature and expect more clear and simple implementation, I could remove this design. :innocent: After all global replacement is not a difficult thing

File: include/LightGBM/config.h
Comment (suncanghuai) [reply include/LightGBM/config.h:866]: Thanks for your suggestion! I will modify the unclear description and fix the failure checks later.

**Comment by guolinke** [comment]
Overall, it seems this change does not affect the efficiency of the normal prediction. So I am okay with the change. 
However, I also agree with @jameslamb , that this new function will increase the future maintain cost.
A workaround may be to create a standalone project that is based on tree inference tools like https://github.com/dmlc/treelite ?

**Comment by suncanghuai** [comment]
> Overall, it seems this change does not affect the efficiency of the normal prediction. So I am okay with the change. However, I also agree with @jameslamb , that this new function will increase the future maintain cost. A workaround may be to create a standalone project that is based on tree inference tools like https://github.com/dmlc/treelite ?

It seems you guys are on the same page about whether this feature should be added to LightGBM.  I'm fine with this decision, at least no more thinking about test cases for this feature saves me much time so that I could find another more worthy issue to work on 😄 If no more discussion, I will close this PR a few hours later.

**Comment by jameslamb** [comment]
Thanks very much @suncanghuai . We really appreciate the time you put into this and your interest in contributing to LightGBM!

We just don't have the maintainer resources right now to take this on.

If you tell me what areas of LightGBM you're interested in, I'd be happy to recommend a different feature or bug!

**Pull request closed without merging**
